zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java

```diff
@@ -676,36 +677,50 @@ public void executeVariableInstruction(
-    incidentBehavior.resolveIncidents(elementInstanceKey);
-
-    catchEventBehavior.unsubscribeFromEvents(elementInstanceKey);
-
-    // terminate all child instances if the element is an event subprocess
-    if (elementType == BpmnElementType.EVENT_SUB_PROCESS
-        || elementType == BpmnElementType.SUB_PROCESS
-        || elementType == BpmnElementType.PROCESS
-        || elementType == BpmnElementType.MULTI_INSTANCE_BODY) {
-      elementInstanceState.getChildren(elementInstanceKey).stream()
-          .filter(ElementInstance::canTerminate)
-          .forEach(this::terminateElement);
-    } else if (elementType == BpmnElementType.CALL_ACTIVITY) {
-      final var calledActivityElementInstance =
-          elementInstanceState.getInstance(elementInstance.getCalledChildInstanceKey());
-      if (calledActivityElementInstance != null && calledActivityElementInstance.canTerminate()) {
-        terminateElement(calledActivityElementInstance);
+    final var elementInstancesToTerminate = new Stack<ElementInstance>();
+    final var elementsTerminating = new Stack<ElementInstance>();
+    elementInstancesToTerminate.push(elementInstance);
+
+    while (!elementInstancesToTerminate.isEmpty()) {
+      final var currentElement = elementInstancesToTerminate.pop();
+      final var elementInstanceKey = currentElement.getKey();
+      final var elementInstanceRecord = currentElement.getValue();
+      final BpmnElementType elementType = currentElement.getValue().getBpmnElementType();
+
+      stateWriter.appendFollowUpEvent(
+          elementInstanceKey, ProcessInstanceIntent.ELEMENT_TERMINATING, elementInstanceRecord);
+
+      jobBehavior.cancelJob(currentElement);
+      incidentBehavior.resolveIncidents(elementInstanceKey);
+
+      catchEventBehavior.unsubscribeFromEvents(elementInstanceKey);
+
+      // terminate all child instances if the element is an event subprocess
+      if (elementType == BpmnElementType.EVENT_SUB_PROCESS
+          || elementType == BpmnElementType.SUB_PROCESS
+          || elementType == BpmnElementType.PROCESS
+          || elementType == BpmnElementType.MULTI_INSTANCE_BODY) {
+        elementInstanceState.getChildren(elementInstanceKey).stream()
+            .filter(ElementInstance::canTerminate)
+            .forEach(elementInstancesToTerminate::add);
+      } else if (elementType == BpmnElementType.CALL_ACTIVITY) {
+        final var calledActivityElementInstance =
+            elementInstanceState.getInstance(currentElement.getCalledChildInstanceKey());
+        if (calledActivityElementInstance != null && calledActivityElementInstance.canTerminate()) {
+          elementInstancesToTerminate.add(calledActivityElementInstance);
+        }
       }
+
+      elementsTerminating.push(currentElement);
     }
 
-    stateWriter.appendFollowUpEvent(
-        elementInstanceKey, ProcessInstanceIntent.ELEMENT_TERMINATED, elementInstanceRecord);
+    while (!elementsTerminating.isEmpty()) {
```

Comment:
❓ I'm curious why was the choice to wait till the end to append all `ELEMENT_TERMINATED` events rather than one by one?

Reply:
It's about the order of the events. Before would recursion it would be:

1. Parent `TERMINATING`
2. Child `TERMINATING`
3. Child `TERMINATED`
4. Parent `TERMINATED`

If I wouldn't append all terminated events what would happen now is:

1. Parent `TERMINATING`
2. Parent `TERMINATED`
3. Child `TERMINATING`
4. Child `TERMINATED`

This is an issue, because handling the `TERMINATED` event for the child requires the parent instance to be available in the state. That would no longer be the case since we deleted it from the state with the parent `TERMINATED` event. To resolve this I collected all the element instances that require a `TERMINATED` event to be written in a stack and worked from the smallest child back to the root element instance.